File: AMoAdiOSSdkDemo(Objective-C)/AMoAdiOSSdkDemo(Objective-C)/InfeedViewController.m
//
//  InfeedViewController.m
//  AMoAdiOSSdkDemo(Object-C)
//
//  Created by AMoAd on 2016/09/02.
//  Copyright © 2016年 AMoAd. All rights reserved.
//

#import "InfeedViewController.h"
#import "AMoAdSdk/AMoAdInfeed.h"

@interface UserItem : NSObject
@property (nonatomic,readwrite,assign) NSInteger index;
@property (nonatomic,readwrite,strong) NSURL *iconUrl;
@property (nonatomic,readwrite,copy) NSString *text1;
@property (nonatomic,readwrite,copy) NSString *text2;
@property (nonatomic,readwrite,copy) NSString *text3;
@property (nonatomic,readwrite,strong) AMoAdItem *adItem;
@end
@implementation UserItem
- (instancetype)initWithAMoAd:(AMoAdItem *)adItem {
    self = [super init];
    if (self) {
        self.index = -1;
        self.iconUrl = [NSURL URLWithString:adItem.iconUrl];
        self.text1 = adItem.serviceName;
        self.text2 = adItem.titleShort;
        self.text3 = adItem.titleLong;
        self.adItem = adItem;
    }
    return self;
}
@end

@interface InfeedViewController() <UITableViewDataSource, UITableViewDelegate>
@property NSArray<UserItem *> *items;
@property int itemNo;
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@end
static NSString *const kUserTableViewCell = @"UserTableViewCell";
@implementation InfeedViewController
- (void)viewDidLoad {
    [super viewDidLoad];
    [AMoAdInfeed setNetworkTimeoutMillis:5000];
    [self initTableView];
    [self addItems];
}

- (void)addItems {
    [self addItems:nil];
}

- (void)addItems:(void(^)()) completion {
    [self loadItems:^(NSArray<UserItem *> *newItems) {
        if(newItems.count > 0) {
            self.items = [self.items arrayByAddingObjectsFromArray:newItems];
            [self.tableView reloadData];
        }
        if (completion){
            completion();
        }
    }];
}

- (void)addItemIfNeeded:(NSIndexPath *) indexPath {
    if (indexPath.row >= self.items.count - 1) {
        [self addItems];
    }
}

- (NSArray<UserItem *> *)mergeAd:(NSArray<UserItem *> *) userItems adList:(AMoAdList *) adList {
    NSMutableArray<UserItem *> *merged = [[NSMutableArray alloc] initWithArray:userItems];
    int adIndex = 0;
    for (int i = 0;i < userItems.count;i++) {
        if (i == adList.beginIndex || (i > adList.beginIndex && (i - adList.beginIndex) % adList.interval == 0)) {
            UserItem *item = [[UserItem alloc]initWithAMoAd:adList.ads[adIndex]];
            adIndex += 1;
            [merged insertObject:item atIndex:i];
        }
        if (adIndex >= adList.ads.count) {
            break;
        }
    }
    return merged;
}

- (void)loadItems:(void (^)(NSArray<UserItem *> *items)) completion {
    //use background thread
    dispatch_async(dispatch_queue_create("com.amoad.queue.UserItems", nil), ^{
        NSArray<UserItem *> *newUserItems = [self createItems];
        if(newUserItems.count > 0) {

            [AMoAdInfeed loadWithSid:self.sid completion:^(AMoAdList *adList, AMoAdResult result) {
                NSArray<UserItem *> *mergedItems;
                if (result == AMoAdResultSuccess && adList.ads.count != 0) {
                    mergedItems = [self mergeAd:newUserItems adList:adList];
                } else {
                    mergedItems = newUserItems;
                }
                //user main thread
                dispatch_async(dispatch_get_main_queue(), ^{
                    completion(mergedItems);
                });
            }];
        }
    });
}

- (void)loadImage:(NSURL *)url completion:(void (^)(UIImage *image))completion {
    dispatch_async(dispatch_queue_create("com.amoad.sample.IconLoadQueue", nil), ^{
        NSURLSession *session = [NSURLSession sharedSession];
        NSURLSessionDataTask *dataTask = [session dataTaskWithURL:url completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
            UIImage *image = nil;
            if (!error && response && data) {
                image = [UIImage imageWithData:data];
            }
            dispatch_async(dispatch_get_main_queue(), ^{
                completion(image);
            });
            
        }];
        [dataTask resume];
    });
}

- (NSArray<UserItem *> *)createItems{
    int start = self.itemNo;
    int end = self.itemNo + 10;
    NSMutableArray<UserItem*> *items = [[NSMutableArray<UserItem*> alloc]init];
    for (int i = start; i < end; i++) {
        UserItem *item = [[UserItem alloc]init];
        item.index = i - 1;
        item.text1 = [NSString stringWithFormat:@"タイトル[%d]", i];
        item.text2 = [NSString stringWithFormat:@"本文１[%d]", i];
        item.text3 = [NSString stringWithFormat:@"本文２[%d]", i];
        [items addObject:item];
    }
    self.itemNo = end;
    return items;
}

- (void)initTableView {
    self.items = [[NSArray alloc]init];
    self.tableView.dataSource = self;
    self.tableView.delegate = self;
    [self.tableView registerNib:[UINib nibWithNibName:kUserTableViewCell bundle:nil] forCellReuseIdentifier:kUserTableViewCell];
    
    UIRefreshControl *refreshControl = [[UIRefreshControl alloc]init];
    [refreshControl addTarget:self action:@selector(onRefresh:) forControlEvents:UIControlEventValueChanged];
    [self.tableView addSubview:refreshControl];
}

- (void)onRefresh:(UIRefreshControl *) refreshControl {
    [refreshControl beginRefreshing];
    self.itemNo = 1;
    self.items = [[NSArray alloc]init];
    [self addItems:^{
        [refreshControl endRefreshing];
    }];
    
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)NSIndexPath {
    return 116;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.items.count;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    UserItem *item = self.items[indexPath.row];
    if (item.adItem) {
        [item.adItem onClick];
    }
}
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:kUserTableViewCell forIndexPath:indexPath];
    UIImageView *iconView = (UIImageView *)[cell viewWithTag:10];
    UILabel *label1 = (UILabel *)[cell viewWithTag:20];
    UILabel *label2 = (UILabel *)[cell viewWithTag:30];
    UILabel *label3 = (UILabel *)[cell viewWithTag:40];
    
    // データを表示する
    UserItem *item = self.items[indexPath.row];
    label1.text = item.text1;
    label2.text = item.text2;
    label3.text = item.text3;
    iconView.image = nil;
    
    if (item.adItem) {
        [self loadImage:item.iconUrl completion:^(UIImage *icon) {
            if ([label1.text isEqualToString:item.text1] &&
                [label2.text isEqualToString:item.text2] &&
                [label3.text isEqualToString:item.text3] &&
                iconView.image == nil) {  // リサイクルされていないか確認する
                iconView.image = icon;
            }
        }];
    } else {
        NSString *resoucePath = [[NSBundle mainBundle] pathForResource:@"item" ofType:@"png"];
        iconView.image = [[UIImage alloc]initWithContentsOfFile:resoucePath];
    }

    [AMoAdInfeed setViewabilityTrackingCell:cell adItem:item.adItem];
    [self addItemIfNeeded:indexPath];
    return cell;
}

@end

File: AMoAdiOSSdkDemo/AMoAdiOSSdkDemo/InfeedViewController.swift
//
//  InfeedTableViewController.swift
//  AMoAdiOSSdkDemo
//
//  Created by AMoAd on 2016/08/16.
//  Copyright © 2016年 AMoAd. All rights reserved.
//

import UIKit

class InfeedViewController: UIViewController, UITableViewDataSource, UITableViewDelegate, AMoAdNativeListViewDelegate {
    let tableViewCell = "UserTableViewCell"
    let tag = "ad01"
    var items = [UserItem]()
    var itemNo = 1
    var sid:String!
    @IBOutlet weak var tableView: UITableView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        AMoAdInfeed.setNetworkTimeoutMillis(5000);

        initTableView()
        addItems()
    }
    
    func initTableView(){
        tableView.delegate = self
        tableView.dataSource = self
        tableView.register(UINib(nibName: tableViewCell, bundle: nil), forCellReuseIdentifier: tableViewCell)
        
        
        let refreshControl = UIRefreshControl()
        refreshControl.addTarget(self, action: #selector(onRefresh(_:)), for: UIControlEvents.valueChanged)
        tableView.addSubview(refreshControl)
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return items.count
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: tableViewCell, for: indexPath)
        let item = items[(indexPath as NSIndexPath).row]

        let iconView = cell.viewWithTag(10) as! UIImageView
        let label1 = cell.viewWithTag(11) as! UILabel
        let label2 = cell.viewWithTag(12) as! UILabel
        let label3 = cell.viewWithTag(13) as! UILabel
        
        label1.text = item.text1
        label2.text = item.text2
        label3.text = item.text3
        iconView.image = nil
        
        if item.adItem != nil {
            loadImage(item.iconUrl, completion: {(icon) in
                if item.text1 == label1.text! &&
                   item.text2 == label2.text! &&
                   item.text3 == label3.text! &&
                   iconView.image == nil { // リサイクルされていないか確認する
                    iconView.image = icon;
                }
            })
        } else {
            iconView.image = UIImage(named: "item")
        }
        
        AMoAdInfeed.setViewabilityTrackingCell(cell, adItem: item.adItem)
        addItemsIfNeeded(indexPath)
        return cell
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return CGFloat(116)
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let item = items[(indexPath as NSIndexPath).row]
        if let adItem = item.adItem {
            adItem.onClick()
        }
    }

    func onRefresh(_ refreshControl:UIRefreshControl) {
        refreshControl.beginRefreshing()
        itemNo = 1
        items = []
        addItems({refreshControl.endRefreshing()})
        
    }
    
    func addItemsIfNeeded(_ indexPath:IndexPath) {
        if (indexPath as NSIndexPath).row >= items.count - 1 {
            addItems()
        }
    }
    
    func createItems() -> [UserItem] {
        let start = itemNo
        let end = itemNo + 10
        var items = [UserItem]()
        for n in start..<end {
            let item = UserItem(text1:"タイトル[\(n)]", text2:"本文１[\(n)]", text3:"本文２[\(n)]")
            items.append(item)
        }
        itemNo = end
        return items
    }
    
    func mergeAd(_ userItems:[UserItem], adList:AMoAdList) -> [UserItem]{
        var merged = [UserItem]() + userItems
        var adIndex = 0
        for (i, _) in userItems.enumerated() {
            if i == adList.beginIndex || (i > adList.beginIndex && (i - adList.beginIndex) % adList.interval == 0) {
                let userItem = UserItem(item: adList.ads[adIndex])
                adIndex += 1
                merged.insert(userItem, at: i)
            }
            if adIndex >= adList.ads.count {
                break
            }
        }
        
        return merged
    }

    func addItems(){
        addItems(nil)
    }

    func addItems(_ completion:(()->Void)?){
        loadItem({(newItems) in
            if newItems.count > 0 {
                self.items = self.items + newItems
                self.tableView.reloadData()
            }
            if let added = completion {
                added()
            }
        })
    }
    
    func loadItem(_ completion:@escaping (_ userItems:[UserItem]) -> Void){
        //use background thread
        DispatchQueue(label: "com.amoad.queue.UserItems", attributes: []).async(execute: {
            let newUserItems = self.createItems()
            if newUserItems.count > 0 {
                AMoAdInfeed.load(withSid: self.sid, completion: {(adList, adResult) in
                    var mergedItems:[UserItem]!
                    if adResult == .success && !(adList?.ads.isEmpty)!{
                        mergedItems = self.mergeAd(newUserItems, adList: adList!)
                    }else{
                        mergedItems = newUserItems
                    }
                    
                    //use main thread
                    DispatchQueue.main.async(execute: {
                        completion(mergedItems)
                    })
                })
            }
        })
        
    }

    func loadImage(_ url:URL, completion:@escaping (_ image:UIImage?) -> Void){
        //use background thread
        DispatchQueue(label: "com.amoad.queue.Icon", attributes: []).async(execute: {
            let request = URLRequest(url: url)
            let response: AutoreleasingUnsafeMutablePointer<URLResponse?>? = nil
            var img:UIImage?
            do {
                let data = try NSURLConnection.sendSynchronousRequest(request, returning: response)
                img = UIImage(data: data)
            }catch (let e){
                print("error:\(e)")
            }

            //use main thread
            DispatchQueue.main.async(execute: {
                completion(img)
            })
        })
    }
    
    class UserItem {
        var iconUrl:URL!
        var text1:String!
        var text2:String!
        var text3:String!
        var adItem:AMoAdItem?
        
        init(text1:String, text2:String, text3:String){
            self.text1 = text1;
            self.text2 = text2;
            self.text3 = text3;
        }
        convenience init(item:AMoAdItem){
            self.init(text1: item.serviceName, text2: item.titleShort, text3: item.titleLong);
            iconUrl = URL(string: item.iconUrl)
            adItem = item
        }
    }
}
